src/Microsoft.VisualStudio.Threading.Shared/AsyncBarrier.cs

```diff
@@ -52,21 +50,33 @@ public AsyncBarrier(int participants)
         /// <returns>An awaitable.</returns>
         public Task SignalAndWait()
         {
-            var tcs = new TaskCompletionSource<bool>();
-            this.waiters.Push(tcs);
-            if (Interlocked.Decrement(ref this.remainingParticipants) == 0)
+            lock (this.waiters)
             {
-                this.remainingParticipants = this.participantCount;
-                var waiters = this.waiters;
-                this.waiters = new ConcurrentStack<TaskCompletionSource<bool>>();
+                if (this.waiters.Count + 1 == this.participantCount)
+                {
+                    // This is the last one we were waiting for.
+                    // Unleash everyone that preceded this one.
+                    while (this.waiters.Count > 0)
+                    {
+                        Task.Factory.StartNew(
+                            state => ((TaskCompletionSource<EmptyStruct>)state).SetResult(default(EmptyStruct)),
+                            this.waiters.Pop(),
+                            CancellationToken.None,
+                            TaskCreationOptions.None,
+                            TaskScheduler.Default);
+                    }
 
-                foreach (var waiter in waiters)
+                    // And allow this one to continue immediately.
+                    return TplExtensions.CompletedTask;
```

Comment:
Should we ``` await Task.Yield(); ``` here? Otherwise, there is an inconsistency: all other waiters got their continuation kicked off asynchronously, and the last one will always be running in the same call stack.